symengine/parser.cpp

```diff
@@ -215,46 +228,56 @@ class ExpressionParser
             iter = operator_end[iter];
         }
 
-        if (params.size() == 1)
+        if (params.size() == 1) {
             if (single_arg_functions.find(expr) != single_arg_functions.end())
                 return single_arg_functions[expr](params[0]);
+        }
 
-        if (params.size() == 2)
+        if (params.size() == 2) {
             if (double_arg_functions.find(expr) != double_arg_functions.end())
                 return double_arg_functions[expr](params[0], params[1]);
+        }
 
-        if (multi_arg_functions.find(expr) != multi_arg_functions.end())
+        if (multi_arg_functions.find(expr) != multi_arg_functions.end()) {
             return multi_arg_functions[expr](params);
+        }
 
         return function_symbol(expr, params);
     }
 
     // return a <Basic> by parsing the 'expr' passed from parse_string
-    RCP<const Basic> set_result(const std::string &expr,
-                                const bool &is_not_numeric, const int &num_dots)
+    RCP<const Basic> set_result(const std::string &expr)
     {
         // for handling cases like "-2"
         // expr will be "" in this case, but we must return 0
         if (expr == "")
             return zero;
 
+        const std::string float_regex
+            = "^[-+]?[0-9]*\\.?[0-9]+([eE][-+]?[0-9]+)?$";
+        const std::string integer_regex = "^[-+]?([1-9][0-9]*|0)$";
+        const std::string symbol_regex = "^[a-zA-Z_$][a-zA-Z_$0-9]*$";
```

Comment:
These should be of type `std::regex`, not `std::string`. Also you should make them global objects I think. So that when you call `std::regex_match` or even just the `set_result` function, you don't need to construct those.